File: proj.android_studio/ee-x-firebase-performance/src/main/java/com/ee/firebase/performance/FirebasePerformance.java
package com.ee.firebase.performance;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;

import com.ee.core.IMessageBridge;
import com.ee.core.IPlugin;
import com.ee.core.Logger;
import com.ee.core.internal.Thread;
import com.ee.core.internal.Utils;
import com.google.firebase.perf.metrics.Trace;

import java.util.HashMap;
import java.util.Map;

public class FirebasePerformance implements IPlugin {
    private static final String k__setDataCollectionEnabled = "FirebasePerformance_setDataCollectionEnabled";
    private static final String k__isDataCollectionEnabled = "FirebasePerformance_isDataCollectionEnabled";
    private static final String k__startTrace = "FirebasePerformance_startTrace";
    private static final String k__newTrace = "FirebasePerformance_newTrace";

    private static final Logger _logger = new Logger(FirebasePerformance.class.getName());

    private IMessageBridge _bridge;
    private com.google.firebase.perf.FirebasePerformance _performance;
    private Map<String, FirebasePerformanceTrace> _traces;

    public FirebasePerformance(@NonNull Context context, @NonNull IMessageBridge bridge) {
        Thread.checkMainThread();
        _bridge = bridge;
        _performance = com.google.firebase.perf.FirebasePerformance.getInstance();
        _traces = new HashMap<>();
        registerHandlers();
    }

    @Override
    public void onCreate(@NonNull Activity activity) {
    }

    @Override
    public void onStart() {
    }

    @Override
    public void onStop() {
    }

    @Override
    public void onResume() {
    }

    @Override
    public void onPause() {
    }

    @Override
    public void onDestroy() {
    }

    @Override
    public void destroy() {
        Thread.checkMainThread();
        deregisterHandlers();
        for (String key : _traces.keySet()) {
            _traces.get(key).destroy();
        }
        _traces.clear();
        _traces = null;
        _performance = null;
    }

    @Override
    public boolean onActivityResult(int requestCode, int responseCode, Intent data) {
        return false;
    }

    private void registerHandlers() {
        Thread.checkMainThread();
        _bridge.registerHandler(k__setDataCollectionEnabled, message -> {
            setDataCollectionEnabled(Utils.toBoolean(message));
            return "";
        });

        _bridge.registerHandler(k__isDataCollectionEnabled, message ->
            Utils.toString(isDataCollectionEnabled()));

        _bridge.registerHandler(k__startTrace, message -> {
            String traceName = message;
            return Utils.toString(startTrace(traceName));
        });

        _bridge.registerHandler(k__newTrace, message -> {
            String traceName = message;
            return Utils.toString(newTrace(traceName));
        });
    }

    private void deregisterHandlers() {
        Thread.checkMainThread();
        _bridge.deregisterHandler(k__setDataCollectionEnabled);
        _bridge.deregisterHandler(k__isDataCollectionEnabled);
        _bridge.deregisterHandler(k__newTrace);
        _bridge.deregisterHandler(k__startTrace);
    }

    public void setDataCollectionEnabled(boolean enabled) {
        _performance.setPerformanceCollectionEnabled(enabled);
    }

    public boolean isDataCollectionEnabled() {
        return _performance.isPerformanceCollectionEnabled();
    }

    public boolean startTrace(@NonNull String traceName) {
        if (_traces.containsKey(traceName)) {
            return false;
        }
        Trace trace = _performance.startTrace(traceName);
        _traces.put(traceName, new FirebasePerformanceTrace(_bridge, trace, traceName));
        return true;
    }

    public boolean newTrace(@NonNull String traceName) {
        if (_traces.containsKey(traceName)) {
            return false;
        }
        Trace trace = _performance.newTrace(traceName);
        _traces.put(traceName, new FirebasePerformanceTrace(_bridge, trace, traceName));
        return true;
    }
}

File: proj.android_studio/ee-x-firebase-performance/src/main/java/com/ee/firebase/performance/FirebasePerformance.kt
package com.ee.firebase.performance

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.annotation.AnyThread
import com.ee.core.IMessageBridge
import com.ee.core.IPlugin
import com.ee.core.Logger
import com.ee.core.internal.Utils
import com.ee.core.registerHandler
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.UnstableDefault
import java.util.concurrent.ConcurrentHashMap

@ImplicitReflectionSerializer
@UnstableDefault
class FirebasePerformance(
    private val _bridge: IMessageBridge,
    private val _context: Context,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val _logger = Logger(FirebasePerformance::class.java.name)

        private const val k__setDataCollectionEnabled = "FirebasePerformance_setDataCollectionEnabled"
        private const val k__isDataCollectionEnabled = "FirebasePerformance_isDataCollectionEnabled"
        private const val k__startTrace = "FirebasePerformance_startTrace"
        private const val k__newTrace = "FirebasePerformance_newTrace"
    }

    private val _performance = com.google.firebase.perf.FirebasePerformance.getInstance()
    private val _traces: MutableMap<String, FirebasePerformanceTrace> = ConcurrentHashMap()

    init {
        registerHandlers()
    }

    override fun onCreate(activity: Activity) {}
    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}
    override fun onDestroy() {}

    override fun destroy() {
        deregisterHandlers()
        for (trace in _traces.values) {
            trace.destroy()
        }
        _traces.clear()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(k__setDataCollectionEnabled) { message ->
            isDataCollectionEnabled = Utils.toBoolean(message)
            ""
        }
        _bridge.registerHandler(k__isDataCollectionEnabled) {
            Utils.toString(isDataCollectionEnabled)
        }
        _bridge.registerHandler(k__startTrace) { message ->
            Utils.toString(startTrace(message))
        }
        _bridge.registerHandler(k__newTrace) { message ->
            Utils.toString(newTrace(message))
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(k__setDataCollectionEnabled)
        _bridge.deregisterHandler(k__isDataCollectionEnabled)
        _bridge.deregisterHandler(k__newTrace)
        _bridge.deregisterHandler(k__startTrace)
    }

    private var isDataCollectionEnabled: Boolean
        @AnyThread get() = _performance.isPerformanceCollectionEnabled
        @AnyThread set(value) {
            _performance.isPerformanceCollectionEnabled = value
        }

    @AnyThread
    fun startTrace(traceName: String): Boolean {
        if (_traces.containsKey(traceName)) {
            return false
        }
        val trace = com.google.firebase.perf.FirebasePerformance.startTrace(traceName)
        _traces[traceName] = FirebasePerformanceTrace(_bridge, trace, traceName)
        return true
    }

    @AnyThread
    fun newTrace(traceName: String): Boolean {
        if (_traces.containsKey(traceName)) {
            return false
        }
        val trace = _performance.newTrace(traceName)
        _traces[traceName] = FirebasePerformanceTrace(_bridge, trace, traceName)
        return true
    }
}